File: generate_clues.py
def generate_clues(spec, data):
  rows = []
  sequence = 0
  for y in xrange(spec['height']):
    clue = []
    for x in xrange(spec['width'] + 1):
      if x < spec['width'] and data[y][x]:
        sequence += 1
      elif sequence > 0:
        clue.append(sequence)
        sequence = 0
    rows.append(clue)

  columns = []
  for x in xrange(spec['width']):
    clue = []
    for y in xrange(spec['height'] + 1):
      if y < spec['height'] and data[y][x]:
        sequence += 1
      elif sequence > 0:
        clue.append(sequence)
        sequence = 0
    columns.append(clue)
  return rows, columns

File: src/client/generate_clues.js
export function generateClues(spec, data) {
  const rows = [];
  let sequence = 0;
  for (let y = 0; y < spec.height; y++) {
    const clue = [];
    for (let x = 0; x <= spec.width; x++) {
      if (x < spec.width && data[y][x]) {
        sequence++;
      } else {
        if (sequence > 0) {
          clue.push(sequence);
          sequence = 0;
        }
      }
    }
    rows.push(clue);
  }
  const columns = [];
  for (let x = 0; x < spec.width; x++) {
    const clue = [];
    for (let y = 0; y <= spec.height; y++) {
      if (y < spec.height && data[y][x]) {
        sequence++;
      } else {
        if (sequence > 0) {
          clue.push(sequence);
          sequence = 0;
        }
      }
    }
    columns.push(clue);
  }
  return [rows, columns];
}